File: src/main/java/org/javaopen/kotlin/updates/Animal.java
package org.javaopen.kotlin.updates;

public class Animal {
    public Animal(String name) {
        setName(name);
    }

    private String name;
    public String getName() { 
        return name;
    }
    public void setName(String name) {
        this.name = name;
    }

    public void action(Animal animal) {
        if (animal instanceof Cat) {
            Cat cat = (Cat)animal;
            cat.scratch();
        } else if (animal instanceof Dog) {
            Dog dog = (Dog)animal;
            dog.bark();
        }
    }
}

class Cat extends Animal {
    public Cat(String name) {
        super(name);
    }

    public void scratch() {
        System.out.println("Scratch this!");
    }
}

class Dog extends Animal {
    public Dog(String name) {
        super(name);
    }
    
    public void bark() {
        System.out.println("Bow wow!");
    }
}

File: src/main/kotlin/org/javaopen/kotlin/updates/Animal.kt
package org.javaopen.kotlin.updates

open class Animal (val name: String)

class Cat(name: String) : Animal(name) {
    fun scratch() {
        println("Scratch it!")
    }
}

class Dog(name: String) : Animal(name) {
    fun bark() {
        println("Bow wow!")
    }
}

fun action(animal: Animal) {
    when (animal) {
        is Cat -> animal.scratch() // Catであることは自明なのでCatのメソッドが呼べる
        is Dog -> animal.bark() // Dogであることは自明なのでDogのメソッドが呼べる
    }
}

// before Kotlin 1.2
fun nonSmartGreeting(name: String?) {
    if (name != null && name.isNotBlank()) {
        println("Hello, ${name.capitalize()}")
    }
}

// after Kotlin 1.3
fun smartGreeting(name: String?) {
    if (!name.isNullOrBlank()) {
        println("Hello, ${name.capitalize()}")
    }
}
